Add WebXR Hit Test reference docs - pt 1 (#8042)
mdn/content: files/en-us/web/api/xrsession/requesthittestsource/index.md

---
title: XRSession.requestHitTestSource()
slug: Web/API/XRSession/requestHitTestSource
tags:
  - API
  - AR
  - Augmented Reality
  - Experimental
  - Method
  - Reference
  - VR
  - Virtual Reality
  - WebXR
  - WebXR Device API
  - XR
  - XRSession
browser-compat: api.XRSession.requestHitTestSource
---
{{APIRef("WebXR Device API")}}

The **`requestHitTestSource()`** method of the

> {{domxref("XRSession")}} interface returns a {{jsxref("Promise")}} that resolves with an {{domxref("XRHitTestSource")}} object that can be used to obtain hit test results using the {{domxref("XRFrame.getHitTestResults()")}} method.

Reviewer: ```suggestion
{{domxref("XRSession")}} interface returns a {{jsxref("Promise")}} that resolves with an {{domxref("XRHitTestSource")}} object that can be passed to {{domxref("XRFrame.getHitTestResults()")}}.
```

## Syntax

```js
requestHitTestSource(options);
```

### Parameters

- `options`
  - : An object containing configuration options, specifically:
    - `space`: The {{domxref("XRSpace")}} that will be tracked by the hit test source.
    - `entityTypes`: {{optional_inline}} An {{jsxref("Array")}} specifying the types of entities that should be used for the purposes of hit test source creation. If no no entity type is specified, the array defaults to a single element with the `plane` type. Possible types:

>       - `point`: Compute hit test results based on characteristic points detected.

Author: hm, yeah, I think this is rather vague and "characteristic points" is what the spec says. Not sure what else to say, though. I believe plane detection and detection using meshes are known concepts to XR/3D developers. For points, this issue argues that point detection could be subsumed into mesh detection: https://github.com/immersive-web/hit-test/issues/66
I'm not expert enough to clarify "points" further. Does it refer to some new detection technique? LIDAR points? Something else? I guess this could be very device-specific also, so maybe that's why the spec left it as such?

Also, the whole `entityTypes` thing is rather unstable still. See https://github.com/immersive-web/hit-test/issues/67 where some people also recognized that this API is rather hard to grok.

>       - `point`: Compute hit test results based on characteristic points detected.

Author: At least I've removed `entityTypes` from the code example now (it is optional anyway).

      - `plane`: Compute hit test results based on real-world planes detected.

>       - `mesh`: Compute hit test results based on meshes detected.

Author: See comment above about this whole thing.

    - `offsetRay`: {{optional_inline}} The {{domxref("XRRay")}} object that will be used to perform hit test. If no `XRRay` object has been provided, a new `XRRay` object is constructed without any parameters.

### Return value

A {{jsxref("Promise")}} that resolves with an {{domxref("XRHitTestSource")}} object.

### Exceptions

Rather than throwing true exceptions, `requestHitTestSource()` rejects the
returned promise with a {{domxref("DOMException")}}, specifically, one of the following:

- `NotSupportedError`

>   - : If `hit-test` is not an enabled feature in {{domxref("XRSystem.requestSession()")}}.

Reviewer: Would 'When' be better than 'If'?

Author: I believe 'If' is better here.

- `InvalidStateError`
  - : If the session has already ended.
- `NotAllowedError`

>   - : If there is an unreasonable amount of requests (some user agents might limit usage due to privacy reasons).

Reviewer: ```suggestion
  - : If there is an unreasonable amount of requests. Some user agents might limit usage for privacy reasons.
```

## Examples

### Requesting a hit test source

> To request a hit test source, make sure to start an {{domxref("XRSession")}} with the `hit-test` session feature enabled. Then you can configure the hit test source and store it for later use in the frame loop where you can use the {{domxref("XRFrame.getHitTestResults()")}} method to obtain the hit test result.

Reviewer: ```suggestion
To request a hit test source, start an {{domxref("XRSession")}} with the `hit-test` session feature enabled. Next, configure the hit test source and store it for later use in the frame loop and call {{domxref("XRFrame.getHitTestResults()")}} to obtain the result.
```
